### PR TITLE
Support Cppcheck

### PR DESCRIPTION
**Pull request by ybiquitous**
- Website: https://github.com/danmar/cppcheck
- Manual: https://github.com/danmar/cppcheck/blob/master/man/manual.md

### CODE REVIEW

**Comment by ybiquitous** [comment]
Oh, v1.89 has been released yesterday... 😓 
https://github.com/danmar/cppcheck/releases/tag/1.89

**Comment by ybiquitous** [comment]
An upgrade is done. Could you please review this? @yykamei

File: images/cppcheck/Dockerfile.erb [hunk 1]
```diff
@@ -0,0 +1,21 @@
...
+
+ARG CPPCHECK_VERSION=1.89
+
+USER root
+RUN cd /tmp \
+  && curl -sSL https://github.com/danmar/cppcheck/archive/${CPPCHECK_VERSION}.tar.gz \
+  |  tar -xz \
+  && cd cppcheck-${CPPCHECK_VERSION} \
+  && make -s MATCHCOMPILER=yes FILESDIR=/usr/share/cppcheck HAVE_RULES=yes CXXFLAGS="-O2 -DNDEBUG -Wall -Wno-sign-compare -Wno-unused-function" \
+  && make -s install MATCHCOMPILER=yes FILESDIR=/usr/share/cppcheck \
```
Comment (yykamei) [inline images/cppcheck/Dockerfile.erb]: As I see `Makefile`, `install` target depends on `cppcehck` target.
So,  it seems sufficient to just append `install` in the [above line](https://github.com/sider/runners/pull/71/files#diff-e1894c83a1521f705f24757ba047b7edR16).

Comment (ybiquitous) [reply images/cppcheck/Dockerfile.erb]: Thank you!

**Comment by ybiquitous** [comment]
One question: Cppcheck compile outputs the followin warnings, but are these no problem?

```
cli/filelister.cpp: In function ‘void addFiles2(std::map<std::__cxx11::basic_string<char>, long unsigned int>&, const string&, const std::set<std::__cxx11::basic_string<char> >&, bool, const PathMatch&)’:
cli/filelister.cpp:197:55: warning: ‘int readdir_r(DIR*, dirent*, dirent**)’ is deprecated [-Wdeprecated-declarations]
             while ((readdir_r(dir, &entry, &dir_result) == 0) && (dir_result != nullptr)) {
                                                       ^
In file included from cli/filelister.cpp:174:
/usr/include/dirent.h:183:12: note: declared here
 extern int readdir_r (DIR *__restrict __dirp,
            ^~~~~~~~~
cli/filelister.cpp:197:55: warning: ‘int readdir_r(DIR*, dirent*, dirent**)’ is deprecated [-Wdeprecated-declarations]
             while ((readdir_r(dir, &entry, &dir_result) == 0) && (dir_result != nullptr)) {
                                                       ^
In file included from cli/filelister.cpp:174:
/usr/include/dirent.h:183:12: note: declared here
 extern int readdir_r (DIR *__restrict __dirp,
            ^~~~~~~~~
```

File: images/cppcheck/Dockerfile.erb (hunk 1 above)
Comment (ybiquitous) [reply images/cppcheck/Dockerfile.erb]: Fixed.

**Comment by yykamei** [comment]
Related PR is [here](https://github.com/danmar/cppcheck/pull/836).
This is not controllable to us so far.

File: lib/node_harness/runners/cppcheck/processor.rb [hunk 2]
```diff
@@ -0,0 +1,126 @@
...
+            parse_result(stderr) do |issue|
+              result.add_issue issue
+            end
+          end
+        end
+
+        def parse_result(output)
+          doc = Nokogiri::XML(output)
+
+          doc.errors.map(&:message).join("\n").tap do |msg|
```
Comment (yykamei) [inline lib/node_harness/runners/cppcheck/processor.rb]: I ran `cppcheck` and got the following result. Using `nokogiri`, `doc.errors` seems empty... What does this block write to stdout?

```xml
<?xml version="1.0" encoding="UTF-8"?>
<results version="2">
    <cppcheck version="1.86"/>
    <errors>
        <error id="nullPointer" severity="warning" msg="Possible null pointer dereference: p" verbose="Possible null pointer dereference: p" cwe="476">
            <location file="ok.cpp" line="4" info="Null pointer dereference"/>
            <location file="ok.cpp" line="20" info="Calling function &apos;f&apos;, 1st argument &apos;p&apos; value is 0"/>
            <location file="ok.cpp" line="19" info="Assignment &apos;p=0&apos;, assigned value is 0"/>
            <symbol>p</symbol>
        </error>
        <error id="eraseDereference" severity="error" msg="Iterator &apos;iter&apos; used after element has been erased." verbose="The iterator &apos;iter&apo
s; is invalid after the element it pointed to has been erased. Dereferencing or comparing it with another iterator is invalid operation." cwe="664">
            <location file="ok.cpp" line="16"/>
            <location file="ok.cpp" line="14"/>
            <symbol>iter</symbol>
        </error>
    </errors>
</results>
```

Comment (ybiquitous) [reply lib/node_harness/runners/cppcheck/processor.rb]: `doc.errors` is not empty when the input XML itself has syntax error(s):

```
$ bundle exec ruby -e 'require "nokogiri"; Nokogiri::XML("<foo>\n<foo>").errors.each{|e| puts e}'
2:6: FATAL: Premature end of data in tag foo line 2
2:6: FATAL: Premature end of data in tag foo line 1
```

See also https://www.rubydoc.info/github/sparklemotion/nokogiri/Nokogiri/XML/Document#errors-instance_method

Comment (yykamei) [reply lib/node_harness/runners/cppcheck/processor.rb]: Thank you, I understood. This is written when an XML is invalid.

Comment (yykamei) [reply lib/node_harness/runners/cppcheck/processor.rb]: :question: In such a case, is it ok to return `NodeHarness::Results::Success`?

Comment (ybiquitous) [reply lib/node_harness/runners/cppcheck/processor.rb]: You are right. 👍 I'll fix.

Comment (ybiquitous) [reply lib/node_harness/runners/cppcheck/processor.rb]: Done.

**Comment by yykamei** [comment]
By the way, the error of `--std` option is confusing to me...

```shell-session
$  cppcheck --std=invalid .
cppcheck: error: unrecognized command line option: "--std=invalid".
```

**Comment by ybiquitous** [comment]
Me too... 😓 

```
$ cppcheck --std posix .
cppcheck: error: unrecognized command line option: "--std".
```

**Pull request merged**